### PR TITLE
Basic double buffer for cpp reader

### LINKED ISSUE #8847: Double buffer for C++ data feeding

### PR DESCRIPTION
**Pull request by JiayiFeng**
fixes #8847 

Please review #8913 first.

### CODE REVIEW

File: paddle/fluid/operators/reader/create_double_buffer_reader_op.cc [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    TensorCopy(buffer_[read_pos_][i], platform::CPUPlace(), &dst);
+    dst.set_lod(buffer_[read_pos_][i].lod());
+    out->push_back(dst);
+  }
+
+  ++read_pos_;
+  if (read_pos_ >= kDoubleBufferSize) {
+    read_pos_ = 0;
+  }
+  buffer_not_full_.notify_all();
```
Comment (dzhwinter) [inline paddle/fluid/operators/reader/create_double_buffer_reader_op.cc]: should be notify_one?

File: paddle/fluid/operators/reader/create_double_buffer_reader_op.cc [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    prefetch.detach();
+  }
+
+  void ReadNext(std::vector<framework::LoDTensor>* out) override;
+  bool HasNext() const override;
+
+ private:
+  void PrefetchThreadFunc();
+
+  std::vector<std::vector<framework::LoDTensor>> buffer_;
```
Comment (dzhwinter) [inline paddle/fluid/operators/reader/create_double_buffer_reader_op.cc]: Maybe a thread safe channel (queue) is better here.

File: python/paddle/fluid/tests/test_cpp_reader.py [hunk 3]
```diff
@@ -15,16 +15,30 @@
 import paddle.v2 as paddle
 import paddle.fluid as fluid
 import numpy as np
+import sys
 
-prog = fluid.framework.Program()
-block = prog.current_block()
+startup_prog = fluid.framework.Program()
```
Comment (dzhwinter) [inline python/paddle/fluid/tests/test_cpp_reader.py]: Why make the reader operator as a part of startup program?

Comment (JiayiFeng) [reply python/paddle/fluid/tests/test_cpp_reader.py]: Because a reader should only be created once.

**Pull request merged**